tests/validate_license_expr.rs

```diff
@@ -0,0 +1,72 @@
+extern crate license_exprs;
+
+use license_exprs::validate_license_expr;
+
+macro_rules! assert_licenses_ok {
+    ($($license:expr),+ $(,)+) => {{
+        $(assert!(validate_license_expr($license).is_ok());)+
+    }}
+}
+
+macro_rules! assert_licenses_err {
+    ($($license:expr),+ $(,)+) => {{
+        $(assert!(validate_license_expr($license).is_err());)+
+    }}
+}
+
+#[test]
+fn single_license() {
+    assert_licenses_ok! {
+        "MIT",
+        "MIT ",
+        " MIT",
+        " MIT ",
+        "    MIT    ",
+        "MIT+",
+    }
+}
+
+#[test]
+fn compound_license() {
+    assert_licenses_ok! {
+        "GPL-3.0+ WITH Classpath-exception-2.0 OR MIT AND AAL",
+        "MIT AND Apache-2.0",
+        "MIT OR Apache-2.0",
+        "GPL-3.0+ WITH Classpath-exception-2.0",
+        "MIT   AND    Apache-2.0",
+    }
+}
+
+#[test]
+fn fails_invalid_license() {
+    assert_licenses_err! {
+        "asdfghjkl",
+        "MIT AND qwerty",
+    }
+}
+
+#[test]
+fn fails_incorrect_structure() {
+    assert_licenses_err! {
+        "()",
+        "(MIT",
+        "MIT)",
+        "MIT Apache-2.0",
+        "MIT and Apache-2.0",
+        "MIT or Apache-2.0",
+        "GPL-3.0+ with Classpath-exception-2.0",
+        "MIT (MIT AND MIT)",
+        "(MIT AND MIT) MIT",
+        "and Apache-2.0",
+        "(MIT and Apache-2.0 and  )",
+        "MIT xor Apache-2.0",
+        "WITH",
+        "MIT OR WITH",
+        // "MIT WITH", // TODO: Incorrectly marked as valid
+        // "MIT AND", // TODO: Incorrectly marked as valid
```

Comment:
Is there a Rust analog to [TAP's TODO tests][1] for marking expected test failures?  It would be nice to report these known failures as non-fatal warnings in the test results and be prompted to remove the TODO once we fix the backing implementation.

[1]: https://testanything.org/tap-version-13-specification.html#todo-tests

Reply:
Not in the default test runner that I know of. I'm hopefully going to fix these errors soon when I work on parsing to an actual syntax tree, which will also involve adding a whole bunch more test cases, so I think keeping them as just comments for now should be ok.